go/libraries/doltcore/sqle/mergeable_indexes_test.go

```diff
@@ -0,0 +1,938 @@
+			2,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15",
+			2,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 15)),
+			},
+			[]int64{6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 OR v1 > 15 OR v1 > 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 OR v1 > 15 AND v1 > 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15 AND v1 > 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 19)),
+			},
+			[]int64{},
+		},
+		{
+			"v1 > 11 OR v1 >= 15",
+			2,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 >= 15",
+			2,
+			[]lookup.Range{
+				lookup.GreaterOrEqualRange(idxv1.tuple(false, 15)),
+			},
+			[]int64{5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 OR v1 > 15 OR v1 >= 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 OR v1 > 15 AND v1 >= 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15 OR v1 >= 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 15)),
+			},
+			[]int64{6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15 AND v1 >= 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterOrEqualRange(idxv1.tuple(false, 19)),
+			},
+			[]int64{9},
+		},
+		{
+			"v1 > 11 OR v1 < 15",
+			2,
+			[]lookup.Range{
+				lookup.AllRange(),
+			},
+			[]int64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 < 15",
+			2,
+			[]lookup.Range{
+				lookup.OpenRange(idxv1.tuple(true, 11), idxv1.tuple(false, 15)),
+			},
+			[]int64{2, 3, 4},
+		},
+		{
+			"v1 > 11 OR v1 > 15 OR v1 < 19",
+			4,
+			[]lookup.Range{
+				lookup.AllRange(),
+			},
+			[]int64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 OR v1 > 15 AND v1 < 19",
+			4,
+			[]lookup.Range{
+				lookup.GreaterThanRange(idxv1.tuple(true, 11)),
+			},
+			[]int64{2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15 OR v1 < 19",
+			4,
+			[]lookup.Range{
+				lookup.AllRange(),
+			},
+			[]int64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 > 15 AND v1 < 19",
+			4,
+			[]lookup.Range{
+				lookup.OpenRange(idxv1.tuple(true, 15), idxv1.tuple(false, 19)),
+			},
+			[]int64{6, 7, 8},
+		},
+		{
+			"v1 > 11 OR v1 <= 15",
+			2,
+			[]lookup.Range{
+				lookup.AllRange(),
+			},
+			[]int64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			"v1 > 11 AND v1 <= 15",
+			2,
+			[]lookup.Range{
+				lookup.CustomRange(idxv1.tuple(true, 11), idxv1.tuple(true, 15),
```

Comment:
This abstraction is bugging me. The CustomRange method should figure out how to deliver a ranged iterator that delivers on its promise. The fact that the caller has to construct tuples that sometimes includes a max ID to get correct behavior is a sign that this abstraction is leaking in a gross way. Think about how you can fix that.

Reply:
I can just move the max ID portion to the Range functions. I didn't do it as I felt it may break on some tuples, but that's only true if the tuple given has a dangling tag with no value. I'd say that that should never happen, and such a tuple is an error/bug so it's fine if it causes more issues further down the stack.

Reply:
Your call, but I don't want the callers of these functions to have to add MAX_TAG artifacts into their tuples. Probably multiple ways to achieve that.